cmd/serve.go

```diff
@@ -1,17 +1,221 @@
+	var cleanUpProxies []func(ctx context.Context) error
+	cleanUpFunc, cleanUpProxies, err := startProxy(logger, appConfig, ctx, cleanUpFunc, cleanUpProxies)
+	if err != nil {
+		return err
+	}
+
+	muxServer := startServer(logger, appConfig, err, ctx, db)
+
+	waitForTermSignal(ctx)
+	cleanup(logger, ctx, cleanUpFunc, cleanUpProxies, muxServer)
+
+	return nil
+}
+
+func cleanup(logger log.Logger, ctx context.Context, cleanUpFunc []func() error, cleanUpProxies []func(ctx context.Context) error, s *server.MuxServer) {
+	for _, f := range cleanUpFunc {
+		if err := f(); err != nil {
+			logger.Warn("error occurred during shutdown", "err", err)
+		}
+	}
+	for _, f := range cleanUpProxies {
+		shutdownCtx, shutdownCancel := context.WithTimeout(ctx, time.Second*20)
+		if err := f(shutdownCtx); err != nil {
+			shutdownCancel()
+			logger.Warn("error occurred during shutdown", "err", err)
+			continue
+		}
+		shutdownCancel()
+	}
+
+	shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer shutdownCancel()
+
+	s.Shutdown(shutdownCtx)
+}
+
+func startServer(logger log.Logger, appConfig *config.Shield, err error, ctx context.Context, db *sql.SQL) *server.MuxServer {
+	s, err := server.NewMux(server.Config{
+		Port: appConfig.App.Port,
+	}, server.WithMuxGRPCServerOptions(getGRPCMiddleware(appConfig, logger)))
+	if err != nil {
+		panic(err)
+	}
+
+	gw, err := server.NewGateway("", appConfig.App.Port)
+	if err != nil {
+		panic(err)
+	}
+
+	handler.Register(ctx, s, gw, apiDependencies(db, appConfig))
+
+	go s.Serve()
+
+	logger.Info("[shield] api is up", "port", appConfig.App.Port)
+
+	// we'll accept graceful shutdowns when quit via SIGINT (Ctrl+C)
+	signal.Notify(proxyTermChan, os.Interrupt, os.Kill, syscall.SIGTERM)
+	return s
+}
+
+func startProxy(logger log.Logger, appConfig *config.Shield, ctx context.Context, cleanUpFunc []func() error, cleanUpProxies []func(ctx context.Context) error) ([]func() error, []func(ctx context.Context) error, error) {
+	for _, service := range appConfig.Proxy.Services {
+		if service.RulesPath == "" {
+			return nil, nil, errors.New("ruleset field cannot be left empty")
+		}
+		blobFS, err := (&blobFactory{}).New(ctx, service.RulesPath, service.RulesPathSecret)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// TODO: option to use default http round tripper for http1.1 backends
+		h2cProxy := proxy.NewH2c(proxy.NewH2cRoundTripper(logger), proxy.NewDirector())
+
+		ruleRepo := blobstore.NewRuleRepository(logger, blobFS)
+		if err := ruleRepo.InitCache(ctx, ruleCacheRefreshDelay); err != nil {
+			return nil, nil, err
+		}
+		cleanUpFunc = append(cleanUpFunc, ruleRepo.Close)
+		pipeline := buildPipeline(logger, h2cProxy, ruleRepo)
+		go func(thisService config.Service, handler http.Handler) {
+			proxyURL := fmt.Sprintf("%s:%d", thisService.Host, thisService.Port)
+			logger.Info("starting h2c proxy", "url", proxyURL)
+
+			//s, err := server.NewMux(server.Config{
```

Comment:
is this comment for todo? if so, can you add `@TODO`? helpful to aggregate todos